[SPARK-27108][SQL] Add parsed SQL plans for create, CTAS.

rdblue opened the pull request:
## What changes were proposed in this pull request?

This moves parsing `CREATE TABLE ... USING` statements into catalyst. Catalyst produces logical plans with the parsed information and those plans are converted to v1 `DataSource` plans in `DataSourceAnalysis`.

This prepares for adding v2 create plans that should receive the information parsed from SQL without being translated to v1 plans first.

This also makes it possible to parse in catalyst instead of breaking the parser across the abstract `AstBuilder` in catalyst and `SparkSqlParser` in core.

For more information, see the [mailing list thread](https://lists.apache.org/thread.html/54f4e1929ceb9a2b0cac7cb058000feb8de5d6c667b2e0950804c613@%3Cdev.spark.apache.org%3E).

## How was this patch tested?

This uses existing tests to catch regressions. This introduces no behavior changes.

rdblue committed:
SPARK-27108: Add parsed SQL plans for create, CTAS.

SparkQA commented:
**[Test build #103224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103224/testReport)** for PR 24029 at commit [`3a77141`](https://github.com/apache/spark/commit/3a771419885b60de6bdfeb5ddaba74391af5ad90).

SparkQA commented:
**[Test build #103224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103224/testReport)** for PR 24029 at commit [`3a77141`](https://github.com/apache/spark/commit/3a771419885b60de6bdfeb5ddaba74391af5ad90).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTable(`
  * `case class CreateTableAsSelect(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103224/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8670/
Test PASSed.

rdblue commented:
@cloud-fan, this is needed to add the v2 create and CTAS plans. We can get a start while waiting for the catalog identifiers to be committed.

rdblue commented:
cc @jzhuge, @mccheah, @gatorsmile

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8671/
Test PASSed.

SparkQA commented:
**[Test build #103225 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103225/testReport)** for PR 24029 at commit [`9bb101f`](https://github.com/apache/spark/commit/9bb101f4541b1a673a184fbcf82f3d4de2d406d8).

SparkQA commented:
**[Test build #103225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103225/testReport)** for PR 24029 at commit [`9bb101f`](https://github.com/apache/spark/commit/9bb101f4541b1a673a184fbcf82f3d4de2d406d8).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTable(`
  * `case class CreateTableAsSelect(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103225/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8685/
Test PASSed.

SparkQA commented:
**[Test build #103240 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103240/testReport)** for PR 24029 at commit [`9e0913f`](https://github.com/apache/spark/commit/9e0913f828a286be37d3adbf09d7a2ff037ace46).

SparkQA commented:
**[Test build #103240 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103240/testReport)** for PR 24029 at commit [`9e0913f`](https://github.com/apache/spark/commit/9e0913f828a286be37d3adbf09d7a2ff037ace46).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class CreateTable(`
  * `case class CreateTableAsSelect(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103240/
Test FAILed.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala:
```diff
@@ -0,0 +1,38 @@
+ * kept in a [[org.apache.spark.sql.catalyst.parser.AbstractSqlParser]].
+ */
+private[sql] abstract class ParsedLogicalPlan extends LogicalPlan {
+  override lazy val resolved = false
```
set this to final at the very least

rdblue replied:
Done.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala:
```diff
@@ -0,0 +1,38 @@
+ * Parsed logical plans are located in Catalyst so that as much SQL parsing logic as possible is be
+ * kept in a [[org.apache.spark.sql.catalyst.parser.AbstractSqlParser]].
+ */
+private[sql] abstract class ParsedLogicalPlan extends LogicalPlan {
```
not sure if this is useful hierarchy, but if yes we should document more clearly this should not survive analysis.

rxin replied:
might be useful to add a special check for this, rather than relying on resolved only.

mccheah replied:
+1 - @rdblue these should only be inputs to the analyzer, not outputs. Would be helpful to write specific JavaDoc on this.

rdblue replied:
That is included above: "Parsed logical plans are not resolved because they must be converted to concrete logical plans."

Do you think that should be rephrase to be more clear?

mccheah replied:
Defer to @rxin but I'm ok with merging with the current docs. We can rephrase in a follow-up if our contributors have trouble with this wording.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala:
```diff
@@ -0,0 +1,64 @@
+import org.apache.spark.sql.types.StructType
+
+/**
+ * A CREATE TABLE command, as parsed from SQL.
```
make it clear here that create table has no data operation. purely metadata.

rdblue replied:
Done.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:
```diff
@@ -376,128 +376,46 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder(conf) {
-   * }}}
+   * TODO: Remove this. It is used because CreateTempViewUsing is not a Catalyst plan.
+   * Either move CreateTempViewUsing into catalyst as a parsed logical plan, or remove it because
+   * it is deprecated.
```
just to double check, when did we deprecate `CreateTempViewUsing`?

rdblue replied:
Looks like it happened in 5effc016c893ce917d535cc1b5026d8e4c846721, 3 years ago.

We can always create a parsed plan for `CreateTempViewUsing` so that it can move to catalyst as well, but I thought that we can do it later, and only need to if this isn't going to be removed in 3.0.

cloud-fan replied:
ah, so we deprecated `CREATE TABLE USING`, but not `CREATE TEMP VIEW USING`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -1860,4 +1861,193 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     if (STRING == null) structField else structField.withComment(string(STRING))
   }
+
+  /**
```
I didn't review this file carefully, assuming it's just copy-paste code from `SparkSqlAstBuilder`

rdblue replied:
Yes, this is moving what is needed from `SparkSqlAstBuilder`. The only real changes are in the `visitCreateTable` method.

These rules should have already been in the abstract class. I'm not sure why they were in `SparkSqlAstBuilder` other than that was the easiest place to put them when they were added.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/ParsedLogicalPlan.scala:
```diff
@@ -0,0 +1,38 @@
+ * converting that v1 metadata to the v2 equivalent, the sql [[CreateTable]] plan is produced by
+ * the parser and converted once into both implementations.
+ *
+ * Parsed logical plans are not resolved because they must be converted to concrete logical plans.
```
is it really necessary to have this parent class just to set the `resolved` bit? I think we can just put `override lazy val resolved = false` in the new `CreateTable` and `CreateTableAsSelect` classes, with classdoc saying that these 2 classes will be replaced by what concrete plans during analysis.

rdblue replied:
The value of this class is that it identifies the set of logical plans that correspond directly to what was parsed from SQL. When someone working on a plan sees `ParsedLogicalPlan` as an ancestor in Scaladoc, it signals what is explained here: that parser produces `ParsedLogicalPlan` nodes without translating what was parsed, then those plans get translated into real plans in the analyzer.

With that information, it is easy to see what changes need to be made. If the parsed plan doesn't include an option, then the parser and parsed plan needs to be updated. If it does include an option, then the analyzer and downstream plans need to be updated.

Also keep in mind that this is the first two subclasses of `ParsedLogicalPlan`. To implement v2 along-side v1, we are going to be adding more of them. So it is valuable that we don't need to remember to set `resolved` to false in every plan.

rdblue committed:
Update for review comments.

* Make ParsedLogicalPlan.resovled final.
* Add docs to CreateTable to be clear that it is metadata-only.

rdblue committed:
Move new DataSource resolution rules into DataSourceResolution.

SparkQA commented:
**[Test build #103461 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103461/testReport)** for PR 24029 at commit [`3e2ca78`](https://github.com/apache/spark/commit/3e2ca7833c98d58006fed50a123ee4e32ee54a69).

rdblue commented:
@cloud-fan, I've responded to the review comments and implemented fixes.

Also, I moved the new resolution rules into `DataSourceResolution`. The new rules needed to be added to `extendedResolutionRules` instead of `postHocResolutionRules` because the post-hoc rules are only run once. This resolution needs to be done before the post-hoc rules defined in `DataSourceAnalysis` run.

At a minimum, these needed to be separated into different classes, but I think it is also more correct for resolution rules to run in the resolution batch so that other resolution rules can run on the plans produced by these rules. This is the same reason why we added the `ResolveOutputRelation` rule to the resolution batch.

I think that the resolution rules in `DataSourceAnalysis` should also move into the resolution batch, but I think that should be done as a separate change.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8879/
Test PASSed.

SparkQA commented:
**[Test build #103461 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103461/testReport)** for PR 24029 at commit [`3e2ca78`](https://github.com/apache/spark/commit/3e2ca7833c98d58006fed50a123ee4e32ee54a69).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with CastSupport  `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103461/
Test FAILed.

rdblue committed:
Fix CreateTableAsSelect conversion to CatalogTable.

SparkQA commented:
**[Test build #103507 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103507/testReport)** for PR 24029 at commit [`0fdfdd4`](https://github.com/apache/spark/commit/0fdfdd4c3b09419ec9bd227c8d0fe3fdb2916a13).

SparkQA commented:
**[Test build #103507 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103507/testReport)** for PR 24029 at commit [`0fdfdd4`](https://github.com/apache/spark/commit/0fdfdd4c3b09419ec9bd227c8d0fe3fdb2916a13).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103507/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8923/
Test PASSed.

rdblue committed:
Apply the v1 source writer list.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8961/
Test PASSed.

SparkQA commented:
**[Test build #103550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103550/testReport)** for PR 24029 at commit [`fc1f41b`](https://github.com/apache/spark/commit/fc1f41bbbbc8e53c5b7a47cc347b248c21ab52cf).

SparkQA commented:
**[Test build #103550 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103550/testReport)** for PR 24029 at commit [`fc1f41b`](https://github.com/apache/spark/commit/fc1f41bbbbc8e53c5b7a47cc347b248c21ab52cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103550/
Test FAILed.

rdblue committed:
Fix V1WriteProvider extraction for Hive sources.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8982/
Test PASSed.

SparkQA commented:
**[Test build #103574 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103574/testReport)** for PR 24029 at commit [`651e5b5`](https://github.com/apache/spark/commit/651e5b580f7132f0f4f87594367ddb0730fa4b4f).

SparkQA commented:
**[Test build #103574 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103574/testReport)** for PR 24029 at commit [`651e5b5`](https://github.com/apache/spark/commit/651e5b580f7132f0f4f87594367ddb0730fa4b4f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103574/
Test FAILed.

rdblue commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8986/
Test PASSed.

SparkQA commented:
**[Test build #103578 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103578/testReport)** for PR 24029 at commit [`651e5b5`](https://github.com/apache/spark/commit/651e5b580f7132f0f4f87594367ddb0730fa4b4f).

rdblue committed:
Fix v1 override check to handle case sensitivity.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8987/
Test PASSed.

SparkQA commented:
**[Test build #103579 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103579/testReport)** for PR 24029 at commit [`7beff42`](https://github.com/apache/spark/commit/7beff421ec1f2896dafa05cc4a1a124adb391039).

SparkQA commented:
**[Test build #103578 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103578/testReport)** for PR 24029 at commit [`651e5b5`](https://github.com/apache/spark/commit/651e5b580f7132f0f4f87594367ddb0730fa4b4f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103578/
Test FAILed.

SparkQA commented:
**[Test build #103579 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103579/testReport)** for PR 24029 at commit [`7beff42`](https://github.com/apache/spark/commit/7beff421ec1f2896dafa05cc4a1a124adb391039).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103579/
Test FAILed.

cloud-fan commented:
retest this please

cloud-fan commented:
LGTM if tests pass

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9020/
Test PASSed.

SparkQA commented:
**[Test build #103619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103619/testReport)** for PR 24029 at commit [`7beff42`](https://github.com/apache/spark/commit/7beff421ec1f2896dafa05cc4a1a124adb391039).

SparkQA commented:
**[Test build #103619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103619/testReport)** for PR 24029 at commit [`7beff42`](https://github.com/apache/spark/commit/7beff421ec1f2896dafa05cc4a1a124adb391039).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103619/
Test FAILed.

rdblue committed:
Fix DDLParserSuite.

* Move CreateTable and CreateTableAsSelect tests to catalyst
* Add PlanResolutionSuite to test parsing and resolution

SparkQA commented:
**[Test build #103637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103637/testReport)** for PR 24029 at commit [`8e7f013`](https://github.com/apache/spark/commit/8e7f0135ff1fdad56f9f5c99bcd29a1ae359678c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9038/
Test PASSed.

rdblue committed:
Fix password redaction tests.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9039/
Test PASSed.

rdblue committed:
Move another test into catalyst DDLParserSuite.

SparkQA commented:
**[Test build #103638 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103638/testReport)** for PR 24029 at commit [`867f2cc`](https://github.com/apache/spark/commit/867f2ccea37381841e4eba852565389fdf0f712f).

SparkQA commented:
**[Test build #103638 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103638/testReport)** for PR 24029 at commit [`867f2cc`](https://github.com/apache/spark/commit/867f2ccea37381841e4eba852565389fdf0f712f).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103638/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9040/
Test PASSed.

SparkQA commented:
**[Test build #103639 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103639/testReport)** for PR 24029 at commit [`51579b6`](https://github.com/apache/spark/commit/51579b6b8128938e75c376283f7a90a4f9e35d65).

SparkQA commented:
**[Test build #103639 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103639/testReport)** for PR 24029 at commit [`51579b6`](https://github.com/apache/spark/commit/51579b6b8128938e75c376283f7a90a4f9e35d65).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103639/
Test FAILed.

rdblue committed:
Fix redaction matcher for type erasure.

SparkQA commented:
**[Test build #103641 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103641/testReport)** for PR 24029 at commit [`65c5cd5`](https://github.com/apache/spark/commit/65c5cd577be4b25c9146c1335663ae95e3f822bd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9042/
Test PASSed.

SparkQA commented:
**[Test build #103637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103637/testReport)** for PR 24029 at commit [`8e7f013`](https://github.com/apache/spark/commit/8e7f0135ff1fdad56f9f5c99bcd29a1ae359678c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DDLParserSuite extends AnalysisTest `
  * `class PlanResolutionSuite extends AnalysisTest `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103637/
Test FAILed.

SparkQA commented:
**[Test build #103641 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103641/testReport)** for PR 24029 at commit [`65c5cd5`](https://github.com/apache/spark/commit/65c5cd577be4b25c9146c1335663ae95e3f822bd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103641/
Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/PlanResolutionSuite.scala:
```diff
@@ -0,0 +1,257 @@
+    }.head
+  }
+
+  test("create table - with partitioned by") {
```
why are some test cases duplicated?

rdblue replied:
These were split into two test suites:

* A DDL suite that validates SQL conversion to the parsed logical plan
* A DDL/Resolution suite that runs the exact same tests to validate that the extra conversion step doesn't change anything

cloud-fan replied:
shall we merge these 2 test suites? Or we need to add document to say that, when adding a new DDL command, we need to add a test in both `DDLParserSuite` and `PlanResolutionSuite` for different purpose.

cloud-fan replied:
shall we merge these 2 test suites? Or we need to add document to say that, when adding a new DDL command, we need to add a test in both `DDLParserSuite` and `PlanResolutionSuite` for different purpose.

rdblue replied:
No.

These suites are in separate modules. The parser suite is in catalyst, where new tests will go. The old tests that validate there are no changes when converting SQL -> parsed plan -> datasource plan must be in core because the plans don't exist in catalyst.

There is no need to add documentation because these will evolve separately. New parsed plans for DDL will be added to the DDL suite. Conversions will be tested when they are added. For example, we are going to add DSv2 plans for create and CTAS. Those conversions will be tested independently.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceResolution.scala:
```diff
@@ -0,0 +1,112 @@
+import org.apache.spark.sql.sources.v2.TableProvider
+import org.apache.spark.sql.types.StructType
+
+case class DataSourceResolution(conf: SQLConf) extends Rule[LogicalPlan] with CastSupport  {
```
shall we call it `DDLResolution`? It's not very related to data source.

rdblue replied:
This resolves parsed plans to `execution.datasources` plans. This isn't just for DDL. We are starting with `CreateTable` and `CreateTableAsSelect`, but there will be more parsed plans that get converted to datasource plans in this rule. That's why I think `DataSourceResolution` is an appropriate name.

mccheah commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/sql/CreateTable.scala:
```diff
@@ -0,0 +1,66 @@
+ *
+ * This is a metadata-only command and is not used to write data to the created table.
+ */
+case class CreateTable(
```
Just so I understand correctly, `DDLResolution` returns an instance of `org.apache.spark.sql.execution.datasources.CreateTable`, which is still a subtype of `LogicalPlan`. It's a little confusing at least to me to have two `LogicalPlan` types be named the same thing, but they represent different steps of the query planning, and are in different packages.

What if we called this variant of `CreateTable`, say, making this `ParsedCreateTable` or perhaps `CreateTableStatement`?

Also, when we start partitioning between V1 sources and V2 sources in the Analyzer, are we going to return the same type of `LogicalPlan` back - a `org.apache.spark.sql.execution.datasources.CreateTable` regardless of it's loading V1 or V2? Or should we instead be naming `org.apache.spark.sql.execution.datasources.CreateTableV1`, anticipating the Analyzer to return instances of `CreateTableV2` plans down the road?

rdblue replied:
I'm okay using packages to differentiate between these. The rules use `sql.CreateTable` and I think that is clear.

I'd also like to avoid renaming the v1 classes to include `V1` in their names. That's going to create a lot of conflicts when working with branches and it's hard enough to maintain PRs already given the amount of needless renaming, whitespace changes, and import changes happening.

Similarly, I want to avoid adding `V2` to class names because we eventually want to remove the old packages and support. Then we would be left with a lot of v2 class names, or we would have to refactor to rename. Using packages as qualifiers is well supported in Scala, so I like to use that for this.

mccheah replied:
In the future then when we do split between `v1` and `v2`, perhaps we can clarify by qualifying just the package? Though I guess that goes against Spark convention, but something like this:

```
import org.apache.spark...v1
import org.apache.spark...v2

...

return v1.CreateTable(...)
...
return v2.CreateTable(...)
```
Otherwise we could alias the imports:

```
import org.apache.spark....v1.{CreateTable => CreateTableV1}
import org.apache.spark....v2.{CreateTable => CreateTableV2}
```

Whatever the case then I think disambiguating by package is fine.

rdblue replied:
Sounds good to me! I'll leave this as is for now.

We can also bring this up in the sync to see if there are other opinions.

cloud-fan replied:
> I'm like using packages to differentiate between the plans.

This reminds me that we add `Exec` postfix to all physical plans, because we think it's not ideal to differentiate plans only be package name. When you search a class by name with an IDE, it will take you more time to pick a class with different package names.

`CreateTableStatemet` sounds good to me.

rdblue committed:
Add Statement suffix to new parsed plans.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9198/
Test PASSed.

rdblue commented:
@mccheah, @cloud-fan, I've rebased on master, fixed a minor conflict, and added the `Statement` suffix to the new plans. I think this should be good to go once tests pass.

SparkQA commented:
**[Test build #103828 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103828/testReport)** for PR 24029 at commit [`5054334`](https://github.com/apache/spark/commit/5054334fdf7f2adfea6f7481f076606c941dd697).

SparkQA commented:
**[Test build #103828 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103828/testReport)** for PR 24029 at commit [`5054334`](https://github.com/apache/spark/commit/5054334fdf7f2adfea6f7481f076606c941dd697).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103828/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9199/
Test PASSed.

SparkQA commented:
**[Test build #103831 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103831/testReport)** for PR 24029 at commit [`6c9b9dc`](https://github.com/apache/spark/commit/6c9b9dc05b815ce9130d5fad7452cda513f57648).

SparkQA commented:
**[Test build #103831 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103831/testReport)** for PR 24029 at commit [`6c9b9dc`](https://github.com/apache/spark/commit/6c9b9dc05b815ce9130d5fad7452cda513f57648).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103831/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

rdblue commented:
Thanks @cloud-fan! And thanks to all the reviewers also!
